pkg/datapath/linux/node_ids.go

```diff
@@ -76,97 +87,180 @@ func (n *linuxNodeHandler) GetNodeID(nodeIP net.IP) (uint16, bool) {
 	localNodeV4 := node.GetIPv4()
 	localNodeV6 := node.GetIPv6()
 	if localNodeV4.Equal(nodeIP) || localNodeV6.Equal(nodeIP) {
-		return 0, true
+		return &localNodeID, true
 	}
 
 	if nodeID, exists := n.nodeIDsByIPs[nodeIP.String()]; exists {
 		return nodeID, true
 	}
 
-	return 0, false
+	return nil, false
+}
+
+func (n *linuxNodeHandler) GetNodeIDForNode(node *nodeTypes.Node) uint16 {
+	nid := n.getNodeIDForNode(node)
+	if nid == nil {
+		return 0
+	}
+	return nid.id
 }
 
 // getNodeIDForNode gets the node ID for the given node if one was allocated
 // for any of the node IP addresses. If none is found, 0 is returned.
-func (n *linuxNodeHandler) getNodeIDForNode(node *nodeTypes.Node) uint16 {
-	nodeID := uint16(0)
+func (n *linuxNodeHandler) getNodeIDForNode(node *nodeTypes.Node) *nodeID {
 	for _, addr := range node.IPAddresses {
-		if id, exists := n.nodeIDsByIPs[addr.IP.String()]; exists {
-			nodeID = id
+		if nodeID, exists := n.nodeIDsByIPs[addr.IP.String()]; exists {
+			return nodeID
 		}
 	}
-	return nodeID
+	return nil
 }
 
-// allocateIDForNode allocates a new ID for the given node if one hasn't already
+func (n *linuxNodeHandler) deleteNodeIPs(oldIPs, newIPs []nodeTypes.Address) {
+	for _, oldAddr := range oldIPs {
+		found := false
+		for _, newAddr := range newIPs {
+			if newAddr.IP.String() == oldAddr.IP.String() {
+				found = true
+				break
+			}
+		}
+		if !found {
+			if err := n.unmapNodeID(oldAddr.IP.String()); err != nil {
+				log.WithError(err).WithFields(logrus.Fields{
+					logfields.IPAddr: oldAddr,
+				}).Warn("DeleteNodeIPS failed to remove a node IP to node ID mapping")
+			}
+		}
+	}
+}
+
+func (n *linuxNodeHandler) allocateIDForNode(node *nodeTypes.Node) uint16 {
+	nodeID := n.allocateNodeID(node)
+	return nodeID.id
+}
+
+// allocateNodeID allocates a new ID for the given node if one hasn't already
 // been assigned. If any of the node IPs have an ID associated, then all other
 // node IPs receive the same. This might happen if we allocated a node ID from
 // the ipcache, where we don't have all node IPs but only one.
-func (n *linuxNodeHandler) allocateIDForNode(node *nodeTypes.Node) uint16 {
+func (n *linuxNodeHandler) allocateNodeID(node *nodeTypes.Node) *nodeID {
 	// Did we already allocate a node ID for any IP of that node?
-	nodeID := n.getNodeIDForNode(node)
+	nID := n.getNodeIDForNode(node)
 
-	if nodeID == 0 {
-		nodeID = uint16(n.nodeIDs.AllocateID())
-		if nodeID == uint16(idpool.NoID) {
+	if nID == nil {
+		id := uint16(n.nodeIDs.AllocateID())
+		if id == uint16(idpool.NoID) {
 			log.WithField(logfields.NodeName, node.Name).Error("No more IDs available for nodes")
 		} else {
 			log.WithFields(logrus.Fields{
-				logfields.NodeID:   nodeID,
+				logfields.NodeID:   id,
 				logfields.NodeName: node.Name,
 			}).Debug("Allocated new node ID for node")
 		}
+		nID = &nodeID{
+			id:     id,
+			refcnt: 1,
+		}
+	} else {
+		nID.refcnt++
```

Comment:
IIUC, we increment the refcount once per node (regardless of number of node IPs) and once per ipcache entry?

Reply:
I thought so to I had the code to do this here but put it in a patch behind this I can add it I agree it does seem more consistent.